design/thin-edge-core.md

```diff
@@ -0,0 +1,50 @@
+# Thin-edge Core Vision
+
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools, actuators, cloud end-points;
+* __constraint environments__ with low-resources devices, large-scale deployments, high exposure to security risks,
+  prohibitive manual operations, long-term supports;
+* __safety related requirements__ with devices operating machines in hazardous contexts.
+  
+The forces being these requirements are pushing in opposite directions.
+
+* On one side, one needs a dynamic and open setting with a large set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, one needs a minimal system specifically built for a task with cherry-picked components.
+  On the extreme side this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel) and sealed, fixed-purpose software images.
+  
+Can these two poles be reconciled?
+
+The approach of thin-edge is to ease the development of IoT applications on edge devices with a smooth transition between:
+* prototyping use-cases - when the aim is to easily explore and combine features,
+* production use-cases - when the need is to deploy a hardened specific software on a large fleet of devices,
+* as well as intermediate use-cases - when the application is build by several vendors with components written using different programming languages;
+  or when a manufacturer assembles his tools on the devices while letting open to his customers the option to add their own modules. 
+
+To achieve this goal, the foundations of Thin-edge address the hardened case first.
+* The core of thin-edge and its components are written using the *Rust* programming languages.
+* The set of components used by an application is defined at *build time*.
+* These components cooperate using a streaming, asynchronous message passing, *internal* API.
+* Connections to the outside are delegated to specific bridge components
+  that abstract the actual protocol (MQTT, HTTP, Modbus, OPC UA)
+  while making these external channels accessible to the other components.
```

Comment:
What does that imply to those "bridge components".
- Do they need to be part of the core?
- Do they need to be defined at build time?
- Do they need to be implemented in Rust?

My expectation for all above would both would be "No, they don't need". Can you confirm?

Reply:
> * Do they need to be part of the core?

No. Even an MQTT connector can be think outside the *core* of thin-edge. An essential plugin, sure, but it can be implemented independently.

> * Do they need to be defined at build time?

Yes. Because establishing a bridge between the core and some outside world.

We can have processes connecting, say OPC UA devices to MQTT. But I would not call these "bridge components" in the context of the thin-edge core.

> * Do they need to be implemented in Rust?

Yes. Because linked into the executable at build time.

Reply:
I rewrote the design principles to make this clearer.